web3_proxy/src/balance.rs

```diff
@@ -0,0 +1,12 @@
+use migration::sea_orm::DatabaseConnection;
+use crate::frontend::errors::FrontendErrorResponse;
+
+/// Implements logic to include balances
+///
+/// Adds an entry to the `spent` table (rpc_requests)
+/// Then subtracts the user balance from the `balance` table
+pub async fn update_balance_and_spend(user_id: u64, db_conn: &DatabaseConnection, amount_delta: u64) -> Result<u64, FrontendErrorResponse> {
```

Comment:
added this as a todo, in case there is more complex / modular logic to be implemented. don't know exactly where to place this yet though (I would assume into the /rpc/ endpoints)

Reply:
This should probably be done in with the code that aggregates and saves data to the rpc_accounting_v2 table. I don't think we need a separate "spent" table. It should be fine to just add a column to rpc_accounting. And it's vital that we aggregate these spends. We do billions of requests and if they aren't aggregated, the db is going to become seriously bloated.

Reply:
I added them here, does this look about right? https://github.com/llamanodes/web3-proxy/pull/35/commits/ea3125e7d4c7392dcc26f8d8d2e76a9ef3bde3ef